### PR TITLE
Implement versioned .js.flow

### PR DESCRIPTION
**Pull request by goodmind**
Proof of concept for #4917 
<s>Has bad performance on all resolutions</s>

## Performance checks

Current measurements: https://gist.github.com/goodmind/65695c3b7a3aa0766812b15779a984dc

1. Build flow binary
2. Run `flow check --profile` on codebase
3. Record results
4. Repeat with regular flow binary and this PR, properly stopping all servers before running new binary

### CODE REVIEW

File: src/parser_utils/package_json.ml [hunk 1]
```diff
@@ -7,18 +7,27 @@
 
 module Ast = Flow_ast
 
+type flow_t = {
+  typesVersions: string SMap.t;
```
Comment (nmote) [inline src/parser_utils/package_json.ml]: Let's name this `types_version` in keeping with typical OCaml naming idioms

File: src/parser_utils/package_json.ml [hunk 2]
```diff
@@ -7,18 +7,27 @@
...
 let (>>=) = Core_result.(>>=)
 
-let empty = { name = None; main = None; }
+let empty = { name = None; main = None; flow = None; }
+
 let name package = package.name
 let main package = package.main
+let flow_types_versions package = match package.flow with
+  | Some t -> t.typesVersions
+  | None -> SMap.empty
```
Comment (nmote) [inline src/parser_utils/package_json.ml]: Is there any case where a client may want to distinguish between a package with no `flow` entry, a package with a `flow` entry that does not contain a `typesVersion` property, and a package with a `flow` entry containing an empty `typesVersion` property?

**Comment by goodmind** [comment]
@nmote how can I test server restart?

File: src/parser_utils/package_json.ml (hunk 2 above)
Comment (goodmind) [reply src/parser_utils/package_json.ml]: I think there isn't any difference, but packages with empty `types_versions` shouldn't be added to cache and skip checking entirely, I think I missed this

Comment (goodmind) [reply src/parser_utils/package_json.ml]: It also depends if we add more fields to `flow`, like #2675

**Comment by nmote** [comment]
I took a quick look and didn't see any existing examples of testing for server restart. With that in mind, I would be happy if you manually test that the server restarts and update the PR description with the steps you took.

**Comment by nmote** [comment]
It appears that we don't have Core's String module available for our internal build, so the build here is failing. I would have mentioned this sooner but this was masked by another build issue that was easy to fix.

@mroch, do you know if it would be reasonable to pull in Core's String? @goodmind, if not, could you rewrite this without using that library?

Sorry, I know it's not a good developer experience to have a separate build system that you can't see or use. Unfortunately there's not a whole lot we can do about that currently.

**Comment by goodmind** [comment]
Yeah, it just needs `substr_index`, couldn't find this in jane street core

**Comment by goodmind** [comment]
@nmote I removed `Base`

**Comment by nmote** [comment]
Unfortunately the test you've added in this commit fails on Linux. Do you have access to a Linux development environment to look into this failure?

**Comment by goodmind** [comment]
@nmote no, I'm on Mac, what fails?

**Comment by goodmind** [comment]
@nmote any news?

**Comment by nmote** [comment]
Yes, I've been trying to get to this. Sorry again for the delay 😞 

I'm planning to rebase it past the ocamlformat change for you, since I'm not sure whether we have ocamlformat properly set up on GitHub yet. I was hoping to get to that yesterday, but ran out of time. I should be able to do that and give you a review today.

**Comment by nmote** [comment]
I've rebased this past the ocamlformat update and I put up #8135 with the updated code. Feel free to take what's there and push to your branch here.

I'll do a round of review over there, so that I don't have to wait for you to update this PR.

**Comment by nmote** [comment]
I've done a round of reviewing over there. Let me know when you get a chance to respond.

**Comment by nmote** [comment]
I'm still concerned about the code in the `types_versions_candidates` function and `node_paths_part.ml` -- see my comment in #8135

**Comment by nmote** [comment]
Closing since there hasn't been any movement on this for some time

**Pull request closed without merging**